feat: exploring bumplot as plotnine geoms

machow opened the pull request:
Hey @JosephBARBIERDARNAL! Don't hesitate to punt this PR, but I've been really interested in trying bumplot, while also working a bit on plotnine extensions. I decided to try out using bumplot from plotnine to see what kind of work is involved.

This is still a WIP, but I implemented a `geom_bezier()` using the bumplot `bezier_curve()` function. I think if you're up for it @has2k1 is down to help with `geom_bezier()` and a bigger `geom_bumplot()`. It could either live here or somewhere else. WDYT?!

Here it is recreating the plot from the README.

```python
import pandas as pd

from plotnine import *
from bumplot.geoms import geom_bezier

data = pd.DataFrame(
    {
        "x": [2020, 2021, 2022, 2023],
        "A": [10, 50, 20, 80],
        "B": [40, 30, 60, 10],
        "C": [90, 20, 70, 40],
    }
)
data_long = data.melt(id_vars="x", value_vars=["A", "B", "C"])

# TODO: I think a plotnine stat could do the ranking
# TODO I just use geom_bezier w/ geom_point, but geom_bumplot could do both
ranked = data_long.assign(
    rank = data_long.groupby("x")["value"].rank(ascending=False).astype(int)
)

(
    ggplot(ranked, aes("x", "rank"))
    # geom_bumplot would essentialy be bezier + point ----
    # plus the rank stat calculation
    + geom_bezier(aes(color="variable"), size=1)
    + geom_point(aes(fill="variable"), size=3, color="black")
    # match readme plot ----
    + theme_minimal()
    + scale_fill_manual(values=["#ffbe0b", "#ff006e", "#3a86ff"])
    + scale_color_manual(values=["#ffbe0b", "#ff006e", "#3a86ff"])
    + scale_y_reverse()
)
```

<img width="1280" height="960" alt="image" src="https://github.com/user-attachments/assets/f685621b-2397-4a6f-8059-ff5600bede1b" />

It seems like this kind of bumplot stuff would be neat to show off in the [2025 Plotnine Plotting Contest](https://github.com/has2k1/plotnine/discussions/categories/2025-plotnine-contest?discussions_q=is%3Aopen+category%3A%222025+Plotnine+Contest%22) 😁

JosephBARBIERDARNAL commented:
Hi @machow

That's a great idea! I hadn't thought of that. In my opinion, it makes sense for `geom_bump()` to be in `bumplot` and think of it as a `plotnine` extension, but I don't have a strong opinion either way.

In [ggbump](https://github.com/davidsjoberg/ggbump), they included a `geom_sigmoid()` function, and I think that makes a little more sense to users than `geom_bezier()` (in terms of the name)?

I don't have much time to think about it right now, but it seems like you already have a pretty good idea on how this could/should be implemented, so if you're willing to work on a more "complete" PR (with @has2k1 ?), I'd be happy to review it, as this would be a very cool addition to `bumplot`!

machow commented:
Alright -- after looking a bit more at ggbump, I've implemented a stat called `stat_rank` to do the ordinal ranking. So basically:

* geom_bump: uses stat rank by default
* geom_bezier: doesn't use rank

RE bezier vs sigmoid: I looked a bit and it seems bezier curves might not use a sigmoid function? I left it as bezier, in case this detail is important. But I didn't look into the bezier code, and am happy to change it if sigmoid seems like a better name!

@has2k1 do you want to take a look at this, in case I messed up any plotnine bits? I added you as a collaborator on the fork, so feel free to push updates!

has2k1 commented on bumplot/geoms.py:
```diff
@@ -0,0 +1,136 @@
+
+    """
+
+    # TODO: I get an error if I don't set geom here,
```
Yes, all stats should have a default geom. The goal of every geom or stat is to create a layer, they simplify creating layers. A layer has both a geom and a stat. That is why geoms have default stats and stats have default geoms.
